### PR TITLE
Create user in `envtool` for PostgreSQL

### LINKED ISSUE #1877: Enforce new authentication
### What should be done?

When new authentication is enabled (by a flag or environment variable), it should be checked at the beginning of every command handler, not in the backend.
When new authentication is not enabled, it should be checked by backends as before.

With #3784 implemented, we should also add integration tests (for new authentication only).

Commands that do not require authentication are:
- `hello`
- `ping`
- `isMaster`
- `whatsmyuri`
- `buildInfo`
- `authenticate`
- `connectionStatus`
- `saslStart` 
- `saslContinue`
- `logout`

### Where?

Handler, backends.

### Definition of Done

- handler updated;
- all backends updated;
- integration/compatibility tests added/updated;
- spot refactorings done;
- **[user documentation](https://docs.ferretdb.io/security/authentication/) updated**.

### PR DESCRIPTION
**Pull request by chilagrow**
# Description

To test it, I added `--test-enable-new-auth` flag in task file `run:` and `run-secured:`.
It does not create user for SQLite, that will be handled in a separate PR.

Closes #1877.

## Readiness checklist

- [x] I added/updated unit tests (and they pass).
- [x] I added/updated integration/compatibility tests (and they pass).
- [ ] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: cmd/envtool/envtool.go [hunk 1]
```diff
@@ -219,6 +222,118 @@ func setupMongodb(ctx context.Context, logger *zap.SugaredLogger) error {
 	return ctx.Err()
 }
 
+// setupUser creates a user in admin database with supported mechanisms.
+// The user uses username/password credential which is the same as the PostgreSQL
+// credentials.
+//
+// Without this, once the first user is created, the authentication fails
+// as username/password does not exist in admin.system.users collection.
+func setupUser(ctx context.Context, logger *zap.SugaredLogger, postgreSQLPort uint16) error {
```
Comment (chilagrow) [inline cmd/envtool/envtool.go:231]: This does not replace [setupUser()](https://github.com/FerretDB/FerretDB/blob/main/integration/setup/setup.go#L334) in integration test.
Because this function creates user in ferretdb database which postgreSQL schema, `postgres://username@127.0.0.1:5432/ferretdb?pool_max_conns=50`.

Each integration test create its own schema like so, `postgres://username@127.0.0.1:5432/testauthentication?pool_max_conns=50`.

File: Taskfile.yml [hunk 2]
```diff
@@ -419,7 +419,7 @@ tasks:
         --proxy-addr=127.0.0.1:47017
         --mode=diff-normal
         --handler=pg
-        --postgresql-url='postgres://127.0.0.1:5433/ferretdb?search_path='
+        --postgresql-url='postgres://username:password@127.0.0.1:5433/ferretdb?search_path='
```
Comment (chilagrow) [inline Taskfile.yml:422]: For run-secured to work, what other better thing can I do?

File: cmd/envtool/envtool.go (hunk 1 above)
Comment (henvic) [reply cmd/envtool/envtool.go:231]: Thanks for the clarification! I was worried this could jeopardize running tests directly with `go test` for a moment 😄

File: integration/commands_authentication_test.go [hunk 3]
```diff
@@ -27,12 +33,55 @@ import (
 func TestCommandsAuthenticationLogout(t *testing.T) {
```
Comment (chilagrow) [inline integration/commands_authentication_test.go:33]: Logout does not fail on authentication check, if user is not authenticated, it just returns success.

File: cmd/envtool/envtool.go (hunk 1 above)
Comment (chilagrow) [reply cmd/envtool/envtool.go]: It's still there for SQLite. It only has handler level authentication, and once the first user is created without this code, auth would fail.

It will be handled in a separate PR.

File: integration/setup/setup.go [hunk 4]
```diff
@@ -334,7 +334,8 @@ func insertBenchmarkProvider(tb testtb.TB, ctx context.Context, collection *mong
 func setupUser(tb testtb.TB, ctx context.Context, client *mongo.Client) {
 	tb.Helper()
 
-	if IsMongoDB(tb) {
+	if IsPostgreSQL(tb) || IsMongoDB(tb) {
+		// the user is created in env tool for PostgreSQL
```
Comment (chilagrow) [inline integration/setup/setup.go:338]: For SQLite we need something similar to have a database with default user. Will be handle in a separate PR.

Comment (henvic) [reply integration/setup/setup.go:338]: What about for MySQL?

**Review by henvic — APPROVED** [review]

File: integration/setup/setup.go (hunk 4 above)
Comment (chilagrow) [reply integration/setup/setup.go:338]: Yes that's true. MySQL hasn't been handled. let me see the status of MySQL backend

File: cmd/envtool/user.json [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      "_id",
+      "credentials",
+      "user",
+      "db",
+      "roles",
+      "userId"
+    ]
+  },
+  "_id": "admin.username",
+  "credentials": {
```
Comment (henvic) [inline cmd/envtool/user.json:121]: Have we decided to use a default username/password?
Does MongoDB do the same? I'm asking because this makes me uncomfortable as it's one of the fragilities most attacked everywhere regarding security.

Comment (chilagrow) [reply cmd/envtool/user.json:121]: This user needs to be the same as the user used in mongodb uri for our test and dev setup to work, for example https://github.com/FerretDB/FerretDB/blob/main/Taskfile.yml#L219]

This env-setup is for our development environment purpose so I'm not concerned with security, we have the same credentials visible in many other places :thinking:

Comment (chilagrow) [reply cmd/envtool/user.json:121]: Or perhaps you mean we should not show hashed value?

Comment (henvic) [reply cmd/envtool/user.json:121]: Got it! Would it make sense to have this named something else like testuser.json? Because from the path alone and no other context I can't know that it's only used for dev / tests. Physically not, though!

**Review by henvic — DISMISSED** [review]

File: cmd/envtool/user.json (hunk 5 above)
Comment (henvic) [reply cmd/envtool/user.json:121]: No. I was wondering about the default value. It doesn't matter if it's hashed or not for me.

**Review by henvic — DISMISSED** [review]

**Review by rumyantseva — APPROVED** [review]
LGTM

File: Taskfile.yml (hunk 2 above)
Comment (AlekSi) [reply Taskfile.yml:422]: That target was supposed to be for the "old" authentication that does not need credentials in the PostgreSQL URI. Why are they needed?

File: cmd/envtool/envtool.go [hunk 6]
```diff
@@ -136,14 +144,46 @@ func setupAnyPostgres(ctx context.Context, logger *zap.SugaredLogger, uri string
...
+	if err != nil {
+		return err
+	}
+
+	err = setupUserInPostgres(ctx, logger, "postgres://username:password@127.0.0.1:5433/ferretdb")
+	if err != nil {
+		return err
+	}
+
+	return setupUserInPostgres(ctx, logger, "postgres://username:password@127.0.0.1:5433/template1")
```
Comment (AlekSi) [inline cmd/envtool/envtool.go:186]: I don't understand why those calls are not inside `setupAnyPostgres` and what we are doing with credentials there

File: cmd/envtool/envtool.go [hunk 7]
```diff
@@ -219,6 +259,92 @@ func setupMongodb(ctx context.Context, logger *zap.SugaredLogger) error {
...
+	defer dbPool.Close()
+
+	var pgErr *pgconn.PgError
+
+	q := `CREATE SCHEMA admin`
+	if _, err = dbPool.Exec(ctx, q); err != nil && (!errors.As(err, &pgErr) || pgErr.Code != pgerrcode.DuplicateSchema) {
+		return err
+	}
+
+	if err == nil {
```
Comment (AlekSi) [inline cmd/envtool/envtool.go:304]: I think we could simply that by:

```go
if errors.As(err, &pgErr) && pgErr.Code == pgerrcode.DuplicateSchema {
  err = nil
}
if err != nil {
  return err
}
```

File: cmd/envtool/envtool.go [hunk 8]
```diff
@@ -219,6 +259,92 @@ func setupMongodb(ctx context.Context, logger *zap.SugaredLogger) error {
...
+		}
+	}
+
+	q = `CREATE TABLE admin.system_users_aff2f7ce (_jsonb jsonb)`
+	if _, err = dbPool.Exec(ctx, q); err != nil && (!errors.As(err, &pgErr) || pgErr.Code != pgerrcode.DuplicateTable) {
+		return err
+	}
+
+	if err == nil {
+		q = `CREATE UNIQUE INDEX "system_users_aff2f7ce__id__67399184_idx" ON admin.system_users_aff2f7ce (((_jsonb->'_id')))`
```
Comment (AlekSi) [inline cmd/envtool/envtool.go:327]: Yikes!..
I wonder if we should step away from that and just implement the localhost exception that we are already working on.

File: integration/setup/setup.go (hunk 4 above)
Comment (AlekSi) [reply integration/setup/setup.go:338]: MySQL and HANA backends are not yet supported, no need to do anything for them

File: cmd/envtool/envtool.go (hunk 8 above)
Comment (chilagrow) [reply cmd/envtool/envtool.go:327]: Yes unfortunately this is something needed to be exposed.
It was copy and paste from log output, logic is too difficult to copy accurately :see_no_evil:

You know what, let me leave this PR for now and progress with local host exception.

File: Taskfile.yml (hunk 2 above)
Comment (chilagrow) [reply Taskfile.yml:422]: It's because postgres started on port 5433 is created from `postgres_secured` docker https://github.com/FerretDB/FerretDB/blob/main/docker-compose.yml#L28.
Unlike `postgres` docker which has `POSTGRES_HOST_AUTH_METHOD=trust`, `postgres_secured` doesn't have `trust` so username/password is required to connect to it.

File: internal/handler/commands.go [hunk 9]
```diff
@@ -203,8 +203,9 @@ func (h *Handler) initCommands() {
 			Help:    "Returns a summary of indexes of the specified collection.",
 		},
 		"logout": {
-			Handler: h.MsgLogout,
-			Help:    "Logs out from the current session.",
+			Handler:   h.MsgLogout,
+			anonymous: true,
```
Comment (AlekSi) [inline internal/handler/commands.go:207]: Let's not forget to port fixes from that PR
